// pastedeno/script/lib/interfaces.js
"use strict";
// Copyright 2023 J.W. Lagendijk. All rights reserved. MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpirationTime = exports.PrivacyLevel = void 0;
var PrivacyLevel;
(function (PrivacyLevel) {
    PrivacyLevel[PrivacyLevel["PUBLIC_ANONYMOUS"] = 0] = "PUBLIC_ANONYMOUS";
    PrivacyLevel[PrivacyLevel["UNLISTED"] = 1] = "UNLISTED";
    PrivacyLevel[PrivacyLevel["PRIVATE"] = 2] = "PRIVATE";
    PrivacyLevel[PrivacyLevel["PUBLIC_USER"] = 3] = "PUBLIC_USER";
})(PrivacyLevel || (exports.PrivacyLevel = PrivacyLevel = {}));
var ExpirationTime;
(function (ExpirationTime) {
    ExpirationTime["NEVER"] = "N";
    ExpirationTime["TEN_MINUTES"] = "10M";
    ExpirationTime["ONE_HOUR"] = "1H";
    ExpirationTime["ONE_DAY"] = "1D";
    ExpirationTime["ONE_WEEK"] = "1W";
    ExpirationTime["TWO_WEEKS"] = "2W";
    ExpirationTime["ONE_MONTH"] = "1M";
    ExpirationTime["SIX_MONTHS"] = "6M";
    ExpirationTime["ONE_YEAR"] = "1Y";
})(ExpirationTime || (exports.ExpirationTime = ExpirationTime = {}));
